lib/mongo/server/connection.rb

```diff
@@ -237,14 +237,26 @@ def handshake!
           min_wire_version = response[Description::MIN_WIRE_VERSION] || Description::LEGACY_WIRE_VERSION
           max_wire_version = response[Description::MAX_WIRE_VERSION] || Description::LEGACY_WIRE_VERSION
           features = Description::Features.new(min_wire_version..max_wire_version)
-          @auth_mechanism = (features.scram_sha_1_enabled? || @server.features.scram_sha_1_enabled?) ? :scram : :mongodb_cr
+
+          if response["ok"] == 1
+            @auth_mechanism = if response['saslSupportedMechs']
```

Comment:
I'm not super fond of the way this logic is expressed and would welcome feedback on it. The basic algorithm is "use SCRAM-SHA-256 if the server responds with it in the list of supported SASL mechanisms, use SCRAM-SHA-1 if the server returns a list of mechanisms without SCRAM-SHA-256 or if no mechanisms are returned by the server but SCRAM-SHA-1 is enabled, and use MONGODB-CR otherwise".

Reply:
I don't see where saslSupportedMechs: "<dbname>.<username>" is being added to the ismaster command. Mechanism negotiation won't work without that.

https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#mechanism-negotiation-via-handshake

Reply:
NM, I found it.